pkg/machineproviders/providers/openshift/machine/v1beta1/provider.go

```diff
@@ -128,16 +167,274 @@ type openshiftMachineProvider struct {
+		Index:        index,
+		ErrorMessage: pointer.StringDeref(machine.Status.ErrorMessage, ""),
+	}, nil
+}
+
+// failureDomainToIndex returns the index of failure domain in the list. If there is nothing found, it returns false as
+// a second parameter.
+func (m *openshiftMachineProvider) failureDomainToIndex(failureDomain failuredomain.FailureDomain) (int32, bool) {
+	for i, fd := range m.indexToFailureDomain {
+		if fd.Equal(failureDomain) {
+			return i, true
+		}
+	}
+
+	return 0, false
+}
+
+// getMachineNameIndex tries to fetch machine index from its name. If it's not possible,
+// it returns false as a second parameter.
+func getMachineNameIndex(machine machinev1beta1.Machine) (int, bool) {
+	// Get a substring after the last occurrence of "-" to find the machine index suffix.
+	// Then try to convert it to a number.
+	machineNameIndex, err := strconv.ParseInt(machine.Name[strings.LastIndex(machine.Name, "-")+1:], 10, 32)
+	if err != nil {
+		return 0, false
+	}
+
+	return int(machineNameIndex), true
+}
+
+// getMachineRef returns returns machine object reference for the given machine.
+func getMachineRef(machine machinev1beta1.Machine) *machineproviders.ObjectRef {
+	return &machineproviders.ObjectRef{
+		ObjectMeta:           machine.ObjectMeta,
+		GroupVersionResource: machinev1beta1.GroupVersion.WithResource("machines"),
+	}
+}
+
+// getNodeRef returns node object reference for the given machine.
+func getNodeRef(machine machinev1beta1.Machine) *machineproviders.ObjectRef {
+	if machine.Status.NodeRef == nil {
+		return nil
+	}
+
+	return &machineproviders.ObjectRef{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: machine.Status.NodeRef.Name,
+		},
+		GroupVersionResource: corev1.SchemeGroupVersion.WithResource("nodes"),
+	}
 }
 
 // CreateMachine creates a new Machine from the template provider config based on the
 // failure domain index provided.
 func (m *openshiftMachineProvider) CreateMachine(ctx context.Context, logger logr.Logger, index int32) error {
+	machineName, err := m.getMachineName(index)
+	if err != nil {
+		return fmt.Errorf("could not generate machine name: %w", err)
+	}
+
+	cpms := &machinev1.ControlPlaneMachineSet{
+		ObjectMeta: m.ownerMetadata,
+	}
+
+	machine := &machinev1beta1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        machineName,
+			Namespace:   m.namespace,
+			Annotations: m.machineTemplate.ObjectMeta.Annotations,
+			Labels:      m.machineTemplate.ObjectMeta.Labels,
+		},
+		Spec: m.machineTemplate.Spec,
+	}
+
+	pc, err := m.providerConfig.InjectFailureDomain(m.indexToFailureDomain[index])
+	if err != nil {
+		return fmt.Errorf("cannot inject failure domain in the provider config: %w", err)
+	}
+
+	rawConfig, err := pc.RawConfig()
+	if err != nil {
+		return fmt.Errorf("cannot fetch raw config from provider config: %w", err)
+	}
+
+	machine.Spec.ProviderSpec.Value.Raw = rawConfig
+
+	if err = controllerutil.SetControllerReference(cpms, machine, m.machineAPIScheme); err != nil {
```

Comment:
Nit, still shadowing
```suggestion
	if err := controllerutil.SetControllerReference(cpms, machine, m.machineAPIScheme); err != nil {
```